src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -1293,84 +1322,197 @@ private void doCancelConnectTimeoutIfNecessary()
+                        }
+
+                        if (canPublish && (reserved != 0 || payloadSize == 0))
+                        {
+                            int willFlags = 0;
+                            if (willFlagSet)
+                            {
+                                if (isSetWillQos(flags))
+                                {
+                                    willFlags = (willFlags | WILL_QOS_MASK) >>> 2;
+                                }
+
+                                if (isSetWillRetain(flags))
+                                {
+                                    willFlags |= RETAIN_FLAG;
+                                }
+                            }
+
+                            resolveSession(sessionStateStream, traceId, authorization, reserved, connect.flags(), willFlags,
+                                payload.expiryInterval, payload.contentType, payload.payloadFormat, payload.responseTopic,
+                                payload.correlationData, sessionPayload, payload.willPayload);
+
+                            doCancelConnectTimeoutIfNecessary();
+                            doEncodeConnack(traceId, authorization, reasonCode, clientIdentifier.value());
+
+                            if (reasonCode == SUCCESS)
+                            {
+                                connected = true;
+                                keepAlive = connect.keepAlive();
+                                keepAliveTimeout = Math.round(TimeUnit.SECONDS.toMillis(keepAlive) * 1.5);
+                                doSignalKeepAliveTimeoutIfNecessary();
+                                decoder = decodePacketType;
+                                progress = connect.limit();
+                            }
+                        }
+                        else
+                        {
+                            this.topicAliasMaximum = -1;
+                            this.sessionExpiryInterval = -1;
+                            decodePublisherKey = topicKey;
+                        }
+                    }
+                    else
+                    {
+                        sessionStateStream = new MqttSessionStateStream(0, topic);
+                    }
+                }
+                else
+                {
+                    if (reasonCode == SUCCESS)
+                    {
+                        doCancelConnectTimeoutIfNecessary();
+                        doEncodeConnack(traceId, authorization, reasonCode, clientIdentifier.value());
+
+                        connected = true;
+                        keepAlive = connect.keepAlive();
+                        keepAliveTimeout = Math.round(TimeUnit.SECONDS.toMillis(keepAlive) * 1.5);
+                        doSignalKeepAliveTimeoutIfNecessary();
+                        decoder = decodePacketType;
+                        progress = connect.limit();
+                    }
```

Comment:
This `else` block duplicates some code above from the `if` block.
Try to bring the codepaths back together and have this code execute less conditionally.